From 7079b27e9cd102994281f62c12d561ee60ad5a22 Mon Sep 17 00:00:00 2001
From: Will Springer <skirmisher@protonmail.com>
Date: Thu, 4 Nov 2021 07:07:45 -0700
Subject: [PATCH] Work around bug in lxc 4.0.10

The bug[1] causes lxc to fail to mount /sys/devices/virtual/net, which
prevents the container from starting. Work around it by replacing `sys.mixed`
with `sys.ro` and adding an explicit rw bind mount, which results in the
original (correct) behavior.

[1]: https://github.com/lxc/lxc/issues/3964
---
 src/anbox/container/lxc_container.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/anbox/container/lxc_container.cpp b/src/anbox/container/lxc_container.cpp
index 87d26b7e8..89a9506ee 100644
--- a/src/anbox/container/lxc_container.cpp
+++ b/src/anbox/container/lxc_container.cpp
@@ -336,7 +336,9 @@ void LxcContainer::start(const Configuration &configuration) {
 
   // We can mount proc/sys as rw here as we will run the container unprivileged
   // in the end
-  set_config_item("lxc.mount.auto", "proc:mixed sys:mixed cgroup:mixed");
+  set_config_item("lxc.mount.auto", "proc:mixed sys:ro cgroup:mixed");
+  // Workaround for https://github.com/lxc/lxc/issues/3964
+  set_config_item("lxc.mount.entry", "/sys/devices/virtual/net sys/devices/virtual/net none rw,bind,optional,create=dir 0 0");
 
   set_config_item("lxc.autodev", "1");
   set_config_item(lxc_config_pty_max_key, "1024");
